File: tour-service/tour-service/src/main/java/com/tourbooking/tourservice/service/WishlistService.java
package com.tourbooking.tourservice.service;

import com.tourbooking.tourservice.exception.BadRequestException;
import com.tourbooking.tourservice.exception.ResourceNotFoundException;
import com.tourbooking.tourservice.model.Tour;
import com.tourbooking.tourservice.model.User;
import com.tourbooking.tourservice.model.Wishlist;
import com.tourbooking.tourservice.repository.TourRepository;
import com.tourbooking.tourservice.repository.UserRepository;
import com.tourbooking.tourservice.repository.WishlistRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
public class WishlistService {

    @Autowired
    private WishlistRepository wishlistRepository;

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private TourRepository tourRepository;

    // Add tour to wishlist
    @Transactional
    public Wishlist addToWishlist(Long userId, Long tourId) {
        // Check user exists
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new ResourceNotFoundException("User not found with id: " + userId));

        // Check tour exists
        Tour tour = tourRepository.findById(tourId)
                .orElseThrow(() -> new ResourceNotFoundException("Tour not found with id: " + tourId));

        // Check if already in wishlist
        if (wishlistRepository.existsByUserIdAndTourId(userId, tourId)) {
            throw new BadRequestException("Tour is already in your wishlist!");
        }

        // Create wishlist item
        Wishlist wishlist = new Wishlist();
        wishlist.setUser(user);
        wishlist.setTour(tour);

        return wishlistRepository.save(wishlist);
    }

    // Remove from wishlist
    @Transactional
    public void removeFromWishlist(Long userId, Long tourId) {
        if (!wishlistRepository.existsByUserIdAndTourId(userId, tourId)) {
            throw new ResourceNotFoundException("Tour not found in wishlist");
        }

        wishlistRepository.deleteByUserIdAndTourId(userId, tourId);
    }

    // Get user's wishlist
    public List<Wishlist> getUserWishlist(Long userId) {
        return wishlistRepository.findByUserId(userId);
    }

    // Check if tour is in wishlist
    public boolean isInWishlist(Long userId, Long tourId) {
        return wishlistRepository.existsByUserIdAndTourId(userId, tourId);
    }

    // Get wishlist count
    public Long getWishlistCount(Long userId) {
        return wishlistRepository.countByUserId(userId);
    }

    // Clear wishlist
    @Transactional
    public void clearWishlist(Long userId) {
        List<Wishlist> items = wishlistRepository.findByUserId(userId);
        wishlistRepository.deleteAll(items);
    }
}

File: tour-booking-frontend/src/services/wishlistService.js
import api from './api';

const wishlistService = {
    // Add to wishlist
    addToWishlist: async (userId, tourId) => {
        try {
            const response = await api.post(`/wishlist?userId=${userId}&tourId=${tourId}`);
            return response.data;
        } catch (error) {
            console.error('Error adding to wishlist:', error);
            throw error;
        }
    },

    // Remove from wishlist
    removeFromWishlist: async (userId, tourId) => {
        try {
            const response = await api.delete(`/wishlist?userId=${userId}&tourId=${tourId}`);
            return response.data;
        } catch (error) {
            console.error('Error removing from wishlist:', error);
            throw error;
        }
    },

    // Get user's wishlist
    getUserWishlist: async (userId) => {
        try {
            const response = await api.get(`/wishlist/user/${userId}`);
            return response.data;
        } catch (error) {
            console.error('Error fetching wishlist:', error);
            throw error;
        }
    },

    // Check if tour is in wishlist
    checkWishlist: async (userId, tourId) => {
        try {
            const response = await api.get(`/wishlist/check?userId=${userId}&tourId=${tourId}`);
            return response.data.isInWishlist;
        } catch (error) {
            console.error('Error checking wishlist:', error);
            return false;
        }
    },

    // Get wishlist count
    getWishlistCount: async (userId) => {
        try {
            const response = await api.get(`/wishlist/count/${userId}`);
            return response.data.count;
        } catch (error) {
            console.error('Error fetching wishlist count:', error);
            return 0;
        }
    },

    // Clear wishlist
    clearWishlist: async (userId) => {
        try {
            const response = await api.delete(`/wishlist/clear/${userId}`);
            return response.data;
        } catch (error) {
            console.error('Error clearing wishlist:', error);
            throw error;
        }
    }
};

export default wishlistService;